# Vuelos de Cabotaje - Argentina

- Modelo
---

In [1]:
# Importamos librerias
import pandas as pd
import numpy as np
from datetime import datetime
from datetime import timedelta

In [2]:
# Configurar pandas para mostrar todas las columnas
pd.set_option('display.max_rows', None)
pd.set_option('display.max_columns', None)

In [3]:
# Cargar a archivos CSV
df_vuelos = pd.read_csv("/content/vuelos_preliminar.csv", low_memory=False, sep=',')

In [4]:
df_vuelos = df_vuelos.drop(df_vuelos.index[-1])

In [5]:
df_vuelos.head(2)

Fecha Fecha_Hora_Local_despegue Tipo_Movimiento Origen Destino  \
0  2021-01-01       2021-01-01 04:09:00        Despegue    EZE     ECA   
1  2021-01-01       2021-01-01 05:10:00        Despegue    EZE     JUJ   

                  Aerolinea          Aeronave  Pasajeros  \
0  AEROLINEAS ARGENTINAS SA       BO-B737-800       70.0   
1  AEROLINEAS ARGENTINAS SA  EMB-ERJ190100IGW       92.0   

                    Ciudad_O   Provincia_O                 Ciudad_D  \
0  EZEIZA/MINISTRO PISTARINI  BUENOS AIRES              EL CALAFATE   
1  EZEIZA/MINISTRO PISTARINI  BUENOS AIRES  JUJUY/GOBERNADOR GUZMÁN   

  Provincia_D                                           Vuelo_ID  \
0  SANTA CRUZ  2021-01-01_EZE_ECA_AEROLINEAS ARGENTINAS SA_BO...   
1       JUJUY  2021-01-01_EZE_JUJ_AEROLINEAS ARGENTINAS SA_EM...   

  Fecha_Hora_Local_aterrizaje Tiempo_Vuelo  
0         2021-01-01 07:11:00     03:02:00  
1         2021-01-01 07:00:00     01:50:00

### Preprocesamiento
- Agregado de nuevos atributos

In [6]:
df_vuelos['temporada_alta'] = 0
df_vuelos['periodo_dia'] = None

In [7]:
# Determinar si el vuelo es de temporada alta
# temporada alta = 1 si Fecha está entre 15-Dic y 3-Mar, o 15-Jul y 31-Jul, o 11-Sep y 30-Sep, 0 si no.

"""def check_condition(date):
    # Convertimos la fecha a datetime y extraemos el mes y el día
    date = pd.to_datetime(date)
    month_day = (date.month, date.day)

    # Definimos los rangos de temporada alta solo con mes y día
    season_ranges = [
        ((12, 15), (12, 31)),  # 15 de diciembre a 31 de diciembre
        ((1, 1), (3, 3)),      # 1 de enero a 3 de marzo
        ((7, 15), (7, 31)),    # 15 de julio a 31 de julio
        ((9, 11), (9, 30))     # 11 de septiembre a 30 de septiembre
    ]

    # Verificamos si la fecha está en cualquiera de los rangos de temporada alta
    for start, end in season_ranges:
        if start <= month_day <= end:
            return 1  # Temporada alta
    return 0  # Fuera de temporada alta

df_vuelos['temporada_alta'] = df_vuelos['Fecha_Hora_Local_despegue'].apply(check_condition)"""

"def check_condition(date):\n    # Convertimos la fecha a datetime y extraemos el mes y el día\n    date = pd.to_datetime(date)\n    month_day = (date.month, date.day)\n\n    # Definimos los rangos de temporada alta solo con mes y día\n    season_ranges = [\n        ((12, 15), (12, 31)),  # 15 de diciembre a 31 de diciembre\n        ((1, 1), (3, 3)),      # 1 de enero a 3 de marzo\n        ((7, 15), (7, 31)),    # 15 de julio a 31 de julio\n        ((9, 11), (9, 30))     # 11 de septiembre a 30 de septiembre\n    ]\n\n    # Verificamos si la fecha está en cualquiera de los rangos de temporada alta\n    for start, end in season_ranges:\n        if start <= month_day <= end:\n            return 1  # Temporada alta\n    return 0  # Fuera de temporada alta\n\ndf_vuelos['temporada_alta'] = df_vuelos['Fecha_Hora_Local_despegue'].apply(check_condition)"

In [8]:
df_vuelos.head(2)

Fecha Fecha_Hora_Local_despegue Tipo_Movimiento Origen Destino  \
0  2021-01-01       2021-01-01 04:09:00        Despegue    EZE     ECA   
1  2021-01-01       2021-01-01 05:10:00        Despegue    EZE     JUJ   

                  Aerolinea          Aeronave  Pasajeros  \
0  AEROLINEAS ARGENTINAS SA       BO-B737-800       70.0   
1  AEROLINEAS ARGENTINAS SA  EMB-ERJ190100IGW       92.0   

                    Ciudad_O   Provincia_O                 Ciudad_D  \
0  EZEIZA/MINISTRO PISTARINI  BUENOS AIRES              EL CALAFATE   
1  EZEIZA/MINISTRO PISTARINI  BUENOS AIRES  JUJUY/GOBERNADOR GUZMÁN   

  Provincia_D                                           Vuelo_ID  \
0  SANTA CRUZ  2021-01-01_EZE_ECA_AEROLINEAS ARGENTINAS SA_BO...   
1       JUJUY  2021-01-01_EZE_JUJ_AEROLINEAS ARGENTINAS SA_EM...   

  Fecha_Hora_Local_aterrizaje Tiempo_Vuelo  temporada_alta periodo_dia  
0         2021-01-01 07:11:00     03:02:00               0        None  
1         2021-01-01 07:00:00     01:50:00               0        None

In [9]:
# Convertimos la columna a formato datetime si no lo está
df_vuelos['Fecha_Hora_Local_despegue'] = pd.to_datetime(df_vuelos['Fecha_Hora_Local_despegue'], errors='coerce')

# Obtenemos el nombre del día de la semana
df_vuelos['dia_nombre'] = df_vuelos['Fecha_Hora_Local_despegue'].dt.day_name()

In [10]:
df_vuelos.head(2)

Fecha Fecha_Hora_Local_despegue Tipo_Movimiento Origen Destino  \
0  2021-01-01       2021-01-01 04:09:00        Despegue    EZE     ECA   
1  2021-01-01       2021-01-01 05:10:00        Despegue    EZE     JUJ   

                  Aerolinea          Aeronave  Pasajeros  \
0  AEROLINEAS ARGENTINAS SA       BO-B737-800       70.0   
1  AEROLINEAS ARGENTINAS SA  EMB-ERJ190100IGW       92.0   

                    Ciudad_O   Provincia_O                 Ciudad_D  \
0  EZEIZA/MINISTRO PISTARINI  BUENOS AIRES              EL CALAFATE   
1  EZEIZA/MINISTRO PISTARINI  BUENOS AIRES  JUJUY/GOBERNADOR GUZMÁN   

  Provincia_D                                           Vuelo_ID  \
0  SANTA CRUZ  2021-01-01_EZE_ECA_AEROLINEAS ARGENTINAS SA_BO...   
1       JUJUY  2021-01-01_EZE_JUJ_AEROLINEAS ARGENTINAS SA_EM...   

  Fecha_Hora_Local_aterrizaje Tiempo_Vuelo  temporada_alta periodo_dia  \
0         2021-01-01 07:11:00     03:02:00               0        None   
1         2021-01-01 07:00:00     01:50:00               0        None   

  dia_nombre  
0     Friday  
1     Friday

In [11]:
# Obtenemos el número del día
df_vuelos['dia_numero'] = df_vuelos['Fecha_Hora_Local_despegue'].dt.day

In [12]:
# Obtenemos el número del mes
df_vuelos['mes_numero'] = df_vuelos['Fecha_Hora_Local_despegue'].dt.month

In [13]:
# Definimos la función para clasificar el periodo del día
# mañana (entre 5:00 y 11:59), tarde (entre 12:00 y 18:59) y noche (entre 19:00 y 4:59), en base a Fecha
def check_condition_2(hms):
    hora = hms.time()  # Extraemos solo la hora de la fecha completa

    if pd.to_datetime('05:00').time() <= hora <= pd.to_datetime('11:59').time():
        return 'mañana'
    elif pd.to_datetime('12:00').time() <= hora <= pd.to_datetime('18:59').time():
        return 'tarde'
    else:
        return 'noche'

# Aplicamos la función a la columna 'Fecha_Hora_Local_despegue'
df_vuelos['periodo_dia'] = df_vuelos['Fecha_Hora_Local_despegue'].apply(check_condition_2)


In [14]:
# Convertimos la columna 'Tiempo_Vuelo' a formato timedelta
df_vuelos['Tiempo_Vuelo'] = pd.to_timedelta(df_vuelos['Tiempo_Vuelo'])

# Luego, obtenemos los minutos
df_vuelos['minutos_vuelo'] = df_vuelos['Tiempo_Vuelo'].dt.total_seconds() / 60

### Seleccion de atributos para el modelo

In [15]:
df_vuelos.head(2)

Fecha Fecha_Hora_Local_despegue Tipo_Movimiento Origen Destino  \
0  2021-01-01       2021-01-01 04:09:00        Despegue    EZE     ECA   
1  2021-01-01       2021-01-01 05:10:00        Despegue    EZE     JUJ   

                  Aerolinea          Aeronave  Pasajeros  \
0  AEROLINEAS ARGENTINAS SA       BO-B737-800       70.0   
1  AEROLINEAS ARGENTINAS SA  EMB-ERJ190100IGW       92.0   

                    Ciudad_O   Provincia_O                 Ciudad_D  \
0  EZEIZA/MINISTRO PISTARINI  BUENOS AIRES              EL CALAFATE   
1  EZEIZA/MINISTRO PISTARINI  BUENOS AIRES  JUJUY/GOBERNADOR GUZMÁN   

  Provincia_D                                           Vuelo_ID  \
0  SANTA CRUZ  2021-01-01_EZE_ECA_AEROLINEAS ARGENTINAS SA_BO...   
1       JUJUY  2021-01-01_EZE_JUJ_AEROLINEAS ARGENTINAS SA_EM...   

  Fecha_Hora_Local_aterrizaje    Tiempo_Vuelo  temporada_alta periodo_dia  \
0         2021-01-01 07:11:00 0 days 03:02:00               0       noche   
1         2021-01-01 07:00:00 0 days 01:50:00               0      mañana   

  dia_nombre  dia_numero  mes_numero  minutos_vuelo  
0     Friday           1           1          182.0  
1     Friday           1           1          110.0

In [16]:
df_raw = df_vuelos[['Origen',	'Destino',	'Aerolinea', 'Aeronave',	'Pasajeros', 	'periodo_dia',	'dia_nombre',	'dia_numero',	'mes_numero', 'minutos_vuelo']].copy()
df_raw.head(2)

Origen Destino                 Aerolinea          Aeronave  Pasajeros  \
0    EZE     ECA  AEROLINEAS ARGENTINAS SA       BO-B737-800       70.0   
1    EZE     JUJ  AEROLINEAS ARGENTINAS SA  EMB-ERJ190100IGW       92.0   

  periodo_dia dia_nombre  dia_numero  mes_numero  minutos_vuelo  
0       noche     Friday           1           1          182.0  
1      mañana     Friday           1           1          110.0

In [17]:
# Convertir las columnas 'dia_numero' y 'mes_numero' a tipo entero
df_raw['dia_numero'] = df_raw['dia_numero'].astype(int)
df_raw['mes_numero'] = df_raw['mes_numero'].astype(int)

df_raw.head(2)

Origen Destino                 Aerolinea          Aeronave  Pasajeros  \
0    EZE     ECA  AEROLINEAS ARGENTINAS SA       BO-B737-800       70.0   
1    EZE     JUJ  AEROLINEAS ARGENTINAS SA  EMB-ERJ190100IGW       92.0   

  periodo_dia dia_nombre  dia_numero  mes_numero  minutos_vuelo  
0       noche     Friday           1           1          182.0  
1      mañana     Friday           1           1          110.0

### Modelo
- Predicción de tiempo de vuelo

In [18]:
%%capture
!pip install catboost

In [19]:
from catboost import CatBoostRegressor
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
from sklearn.metrics import r2_score
pd.options.display.float_format = '{:.4f}'.format

In [20]:
# Eliminamos dos atributos no necesarios para este modelo y la variable dependiente
X1 = df_raw.drop(['Aeronave','Pasajeros','minutos_vuelo'], axis=1)

In [21]:
y1 = df_raw.loc[:, df_raw.columns == "minutos_vuelo" ]

In [22]:
# Dividir el conjunto de datos en entrenamiento y prueba
X_train_1, X_test_1, y_train_1, y_test_1 = train_test_split(X1, y1, test_size=0.2, random_state=42)

In [23]:
# Especificar las columnas categóricas
categorical_features = ['Origen',	'Destino',	'Aerolinea',	'periodo_dia',	'dia_nombre',	'dia_numero',	'mes_numero']

In [24]:
# Crear y entrenar el modelo
model_1 = CatBoostRegressor(cat_features=categorical_features, verbose=0)
model_1.fit(X_train_1, y_train_1)

In [25]:
# Hacer predicciones
y_pred_test_1 = model_1.predict(X_test_1)

In [26]:
# Evaluar el modelo
mse_1 = mean_squared_error(y_test_1, y_pred_test_1)
rmse_1  = pd.to_numeric(mse_1**(1/2), downcast='float')
r2_1 = r2_score(y_test_1 ,y_pred_test_1)
print(f'Mean Squared Error: {round(mse_1)}')
print(f'Root Mean Squared Error: {round(rmse_1)} minutos')
print(f"R2: ", round(r2_1,4))

Mean Squared Error: 203
Root Mean Squared Error: 14 minutos
R2:  0.8642


- Estimacion

In [27]:
Aerolineas = ['AEROLINEAS ARGENTINAS SA','FB LÍNEAS AÉREAS - FLYBONDI','JETSMART AIRLINES S.A.']
Origen = 'EZE'
Destino = 'SAL'
Aerolinea = Aerolineas[1]
periodo_dia = 'tarde'
dia_nombre = 'Monday'
dia_numero = 15
mes_numero = 7

In [28]:
X_entrada = pd.DataFrame({
    'Origen': [Origen],
    'Destino': [Destino],
    'Aerolinea': [Aerolinea],
    'periodo_dia': [periodo_dia],
    'dia_nombre': [dia_nombre],
    'dia_numero': [dia_numero],
    'mes_numero': [mes_numero]
})

In [29]:
y_est_1 = model_1.predict(X_entrada)
t_vuelo = round(y_est_1[0])
print(f"La duracion del vuelo es de: {t_vuelo} minutos")

La duracion del vuelo es de: 111 minutos


In [30]:
df_raw[(df_raw['Origen'] == 'EZE')
       & (df_raw['Destino'] == 'BAR')
       & (df_raw['Aerolinea'] == 'AEROLINEAS ARGENTINAS SA')
       & (df_raw['periodo_dia'] == 'mañana')
       & (df_raw['dia_nombre'] == 'Monday')
       & (df_raw['dia_numero'] == 15)
       ]

Origen Destino                 Aerolinea      Aeronave  Pasajeros  \
2777     EZE     BAR  AEROLINEAS ARGENTINAS SA    BO-737-800   178.0000   
2782     EZE     BAR  AEROLINEAS ARGENTINAS SA  BO-B-737-76N   126.0000   
4422     EZE     BAR  AEROLINEAS ARGENTINAS SA    BO-737-8HX   132.0000   
4432     EZE     BAR  AEROLINEAS ARGENTINAS SA    BO-737-800   168.0000   
21186    EZE     BAR  AEROLINEAS ARGENTINAS SA  AIB-A330-202   210.0000   
21200    EZE     BAR  AEROLINEAS ARGENTINAS SA   BO-B737-8MB   132.0000   

      periodo_dia dia_nombre  dia_numero  mes_numero  minutos_vuelo  
2777       mañana     Monday          15           2       114.0000  
2782       mañana     Monday          15           2       113.0000  
4422       mañana     Monday          15           3       115.0000  
4432       mañana     Monday          15           3       111.0000  
21186      mañana     Monday          15          11       116.0000  
21200      mañana     Monday          15          11       120.0000

In [32]:
df_tiempo_vuelos = pd.read_csv("/content/promedio_tiempos_vuelo.csv", low_memory=False, sep=',')


In [33]:
df_tiempo_vuelos.head()

Aerolinea Origen Destino Tiempo_Promedio
0  AEROLINEAS ARGENTINAS SA    AER     AER        01:29:50
1  AEROLINEAS ARGENTINAS SA    AER     BAR        02:01:53
2  AEROLINEAS ARGENTINAS SA    AER     BCA        01:02:32
3  AEROLINEAS ARGENTINAS SA    AER     CAT        01:35:32
4  AEROLINEAS ARGENTINAS SA    AER     CBA        01:07:35

In [34]:
# Funcion que determina si el tiempo de vuelo presentará una demora o no en base a los datos de df_tiempo_vuelos
#def demora(Aerolinea,Origen,Destino,t_vuelo):
  #obtener tiempo promedio del dataset
tiempo_promedio = df_tiempo_vuelos[(df_tiempo_vuelos['Aerolinea'] == Aerolinea)
                                    & (df_tiempo_vuelos['Origen'] == Origen)
                                    & (df_tiempo_vuelos['Destino'] == Destino)]['Tiempo_Promedio'].values[0]
print(tiempo_promedio)
# Transformarlo a minutos
tiempo_promedio = round((pd.to_timedelta(tiempo_promedio).total_seconds() / 60),2)
tiempo_promedio

01:50:56


110.93

In [35]:
#calcular diferencia entre t_vuelo y tiempo_promedio (modulo)
diferencia = round(abs(t_vuelo - tiempo_promedio))
diferencia

if diferencia < 15: # a partir de 15 min se considera demora
  print(f"No hay demora en el vuelo: {diferencia} minutos")
  demora = 0
else:
  print(f"Hay demora en el vuelo, y es de:{diferencia} minutos")
  demora = 1

No hay demora en el vuelo: 0 minutos


### Modelo
- Predicción de Ocupación de Vuelos

In [ ]:
# Eliminamos variables innecesarias y la variable dependiente
X2 = df_raw.drop(['minutos_vuelo'], axis=1)

In [ ]:
# Especificar las columnas categóricas
categorical_features = ['Origen',	'Destino',	'Aerolinea',	'temporada_alta',	'periodo_dia',	'dia_nombre',	'dia_numero',	'mes_numero']

# Task
Here's a summary of the provided notebook, detailing its purpose, the data preprocessing steps, the flight time prediction model, the use of `promedio_tiempos_vuelo.csv`, and identified areas for improvement:

**General Explanation of the Notebook**
The primary purpose of this notebook is to preprocess flight data and develop a machine learning model to predict flight duration in minutes. Additionally, it aims to identify potential flight delays by comparing predicted flight times with historical average flight times.

**Detailed Data Preprocessing Analysis**
1.  **Data Loading**: The notebook starts by loading flight data from `vuelos_preliminar.csv` into a pandas DataFrame `df_vuelos`.
2.  **Row Removal**: The last row of the `df_vuelos` DataFrame is explicitly dropped, likely to handle potential incomplete or erroneous data at the end of the file.
3.  **Data Type Conversion**:
    *   `Fecha_Hora_Local_despegue` is converted to datetime objects.
    *   `Tiempo_Vuelo` is converted to timedelta objects.
    *   `dia_numero` and `mes_numero` are explicitly converted to integer types.
4.  **New Feature Creation**:
    *   `temporada_alta`: This column is initialized to 0. Although there is commented-out code to calculate it based on specific date ranges, this code is not executed, meaning the column remains all zeros.
    *   `periodo_dia`: Categorizes the time of day (morning, afternoon, night) based on the `Fecha_Hora_Local_despegue`.
    *   `dia_nombre`: Extracts the day name (e.g., 'Monday') from `Fecha_Hora_Local_despegue`.
    *   `dia_numero`: Extracts the day of the month from `Fecha_Hora_Local_despegue`.
    *   `mes_numero`: Extracts the month number from `Fecha_Hora_Local_despegue`.
    *   `minutos_vuelo`: Calculates the total flight duration in minutes from the `Tiempo_Vuelo` timedelta column.
5.  **Attribute Selection for Model**: A new DataFrame `df_raw` is created, selecting relevant columns: `Origen`, `Destino`, `Aerolinea`, `Aeronave`, `Pasajeros`, `periodo_dia`, `dia_nombre`, `dia_numero`, `mes_numero`, and `minutos_vuelo`.

**Flight Time Prediction Model Explanation (`model_1`)**
1.  **Model Type**: A `CatBoostRegressor` model is used for predicting flight duration.
2.  **Input Variables**: The model `model_1` uses `Origen`, `Destino`, `Aerolinea`, `periodo_dia`, `dia_nombre`, `dia_numero`, and `mes_numero` as input features (`X1`). `Aeronave` and `Pasajeros` are explicitly dropped for this model.
3.  **Target Variable**: The target variable `y1` is `minutos_vuelo`.
4.  **Categorical Features**: The model is configured to treat `Origen`, `Destino`, `Aerolinea`, `periodo_dia`, `dia_nombre`, `dia_numero`, and `mes_numero` as categorical features.
5.  **Data Split**: The data is split into training and testing sets with an 80/20 ratio (`test_size=0.2`) and a `random_state=42`.
6.  **Training and Prediction**: The `model_1` is trained on `X_train_1`, `y_train_1` and then used to make predictions (`y_pred_test_1`) on `X_test_1`.
7.  **Evaluation Metrics**: The model's performance is evaluated using:
    *   Mean Squared Error (MSE): 203
    *   Root Mean Squared Error (RMSE): 14 minutes
    *   R-squared (R2): 0.8642
    This indicates a reasonably good fit, with predictions typically within 14 minutes of the actual flight time.
8.  **Prediction Usage**: An example demonstrates how to use the trained `model_1` to predict the flight duration for a new set of input parameters.

**Purpose of `promedio_tiempos_vuelo.csv`**
The `promedio_tiempos_vuelo.csv` file is loaded and utilized to determine if a predicted flight duration signifies a significant delay. The function (or block of code mimicking a function) retrieves the historical average flight time (`Tiempo_Promedio`) for a given airline, origin, and destination from this CSV. It then compares the model's predicted flight time (`t_vuelo`) with this historical average. A difference (absolute value) of 15 minutes or more is classified as a delay.

**Improvements and Surplus Elements**
1.  **Unused `temporada_alta` Calculation**: The code to determine `temporada_alta` based on date ranges (cell `GIYzzxM-Mqvx`) is commented out. As a result, the `temporada_alta` column (initialized to 0) remains unused and contains no meaningful information. This commented code could either be activated or removed if not intended for use.
2.  **Redundant Initialization**: `temporada_alta` and `periodo_dia` are initialized in cell `LPB1OYSUMhSM`, but `periodo_dia` is subsequently overwritten with calculated values. The initial assignment for `temporada_alta` (to 0) is effectively the final state because the calculation is commented out.
3.  **Incomplete Second Model (`X2`)**: There's a section titled "Predicción de Ocupación de Vuelos" where a DataFrame `X2` is created, and categorical features are defined for a potential second model. However, no model is implemented or trained for this task, leaving this section incomplete and `X2` unused.
4.  **Code Clean-up**: Removing the commented-out `temporada_alta` calculation and the incomplete `X2` model section would significantly clean up the notebook and improve readability by focusing only on the active and complete functionalities.

**Final Task**
The notebook successfully preprocesses flight data, creates new features, and trains a CatBoostRegressor model to predict flight duration with an RMSE of 14 minutes and an R2 of 0.8642. It further utilizes a separate CSV (`promedio_tiempos_vuelo.csv`) to determine if a predicted flight duration indicates a delay based on a 15-minute threshold. For improvement, the notebook should either implement the `temporada_alta` calculation or remove the related code, and complete or remove the incomplete second model for flight occupation prediction.

## Explicación General de la Notebook

### Subtask:
Describir el propósito principal de la notebook, que es el preprocesamiento de datos de vuelos y la creación de un modelo para predecir el tiempo de vuelo, así como la detección de posibles demoras.


El objetivo principal de esta notebook es procesar y analizar datos de vuelos para construir un modelo predictivo. Se enfoca en el preprocesamiento de datos de vuelos, la creación de nuevos atributos relevantes y el desarrollo de un modelo capaz de predecir el tiempo de vuelo. Adicionalmente, la notebook incluye la funcionalidad para detectar posibles demoras en los vuelos basándose en comparaciones con tiempos de vuelo promedio históricos, proporcionando así una herramienta integral para el análisis y la predicción de operaciones aéreas.

## Análisis Detallado del Preprocesamiento de Datos

### Subtask:
Explicar cada paso del preprocesamiento, incluyendo la carga de datos, la eliminación de la última fila, la conversión de tipos de datos, la creación de nuevas características como 'temporada_alta', 'periodo_dia', 'dia_nombre', 'dia_numero', 'mes_numero' y 'minutos_vuelo', y la selección de atributos para el modelo.


## Análísis Detallado del Preprocesamiento de Datos

### Carga de Datos
El proceso comienza con la carga del archivo `vuelos_preliminar.csv` en un DataFrame de pandas llamado `df_vuelos`. Este DataFrame es la base para todo el análisis y modelado subsiguiente.

### Eliminación de la Última Fila
Se eliminó la última fila del DataFrame `df_vuelos` (`df_vuelos = df_vuelos.drop(df_vuelos.index[-1])`). Esta es una práctica común en el preprocesamiento de datos, especialmente si la última fila contiene metadatos, resúmenes o datos incompletos que no son relevantes para el análisis.

### Conversión de Tipos de Datos
Se realizaron las siguientes conversiones de tipos de datos para asegurar la correcta manipulación y cálculo de las variables:
*   La columna `Fecha_Hora_Local_despegue` fue convertida al tipo de dato `datetime` para permitir la extracción de componentes de fecha y hora.
*   La columna `Tiempo_Vuelo` fue convertida al tipo `timedelta` para representar duraciones de tiempo. A partir de esta, se calculó una nueva columna `minutos_vuelo`, que representa la duración del vuelo en minutos, dividiendo el total de segundos por 60.
*   Las columnas `dia_numero` y `mes_numero`, extraídas previamente, se convirtieron explícitamente a tipo `int` para un manejo más eficiente y consistente.

### Creación de Nuevas Características
Se enriqueció el DataFrame con nuevas características derivadas de las existentes:
*   **`temporada_alta`**: Esta columna fue inicializada a `0`. Aunque el código para determinar la temporada alta de forma dinámica basado en rangos de fechas fue incluido y comentado, no se ejecutó, por lo que la columna permanece con su valor inicial de `0`.
*   **`periodo_dia`**: Se creó para clasificar los vuelos en 'mañana' (05:00 a 11:59), 'tarde' (12:00 a 18:59) o 'noche' (19:00 a 04:59). Esta clasificación se realizó aplicando la función `check_condition_2` a la hora de despegue (`Fecha_Hora_Local_despegue`).
*   **`dia_nombre`, `dia_numero`, `mes_numero`**: Estas características se extrajeron directamente de la columna `Fecha_Hora_Local_despegue`. `dia_nombre` proporciona el nombre del día de la semana (ej. 'Monday'), mientras que `dia_numero` y `mes_numero` proporcionan el número del día y el mes, respectivamente.

### Selección de Atributos para el Modelo
Para la fase de modelado, se seleccionó un subconjunto de las columnas originales y las características recién creadas para formar el DataFrame `df_raw`. Las columnas elegidas son: 'Origen', 'Destino', 'Aerolinea', 'Aeronave', 'Pasajeros', 'periodo_dia', 'dia_nombre', 'dia_numero', 'mes_numero' y 'minutos_vuelo'. Esta selección se basa en la relevancia de estas variables para la predicción del tiempo de vuelo y la ocupación.

## Explicación del Modelo de Predicción de Tiempo de Vuelo

### Subtask:
Detallar el proceso de creación y entrenamiento del modelo `model_1` (CatBoostRegressor), las variables utilizadas (categóricas), la división de datos (entrenamiento/prueba) y la evaluación de sus métricas (MSE, RMSE, R2). También se explicará cómo se usa para realizar predicciones.


## Explicación del Modelo de Predicción de Tiempo de Vuelo

### Tipo de Modelo
Para la predicción de la duración de los vuelos, se utilizó un modelo **CatBoostRegressor**. CatBoost es un algoritmo de potenciación de gradiente de código abierto, diseñado para ser eficiente y preciso, especialmente con datos categóricos.

### Variables de Entrada y Objetivo
El modelo `model_1` fue entrenado para predecir la duración del vuelo en minutos. Las variables de entrada (características) utilizadas son:
- `Origen`: Código IATA del aeropuerto de origen.
- `Destino`: Código IATA del aeropuerto de destino.
- `Aerolinea`: Nombre de la aerolínea.
- `periodo_dia`: Período del día (mañana, tarde, noche) del despegue.
- `dia_nombre`: Día de la semana del despegue.
- `dia_numero`: Número del día del mes del despegue.
- `mes_numero`: Número del mes del despegue.

Para este modelo, las columnas `Aeronave` y `Pasajeros` fueron descartadas del conjunto de características (`X1`). La variable objetivo (la que se busca predecir) es `minutos_vuelo`.

### Características Categóricas
CatBoost maneja automáticamente las características categóricas. En este modelo, las siguientes columnas fueron especificadas como características categóricas:
- `'Origen'`
- `'Destino'`
- `'Aerolinea'`
- `'periodo_dia'`
- `'dia_nombre'`
- `'dia_numero'`
- `'mes_numero'`

### División de Datos
El conjunto de datos original se dividió en dos subconjuntos: un conjunto de entrenamiento y un conjunto de prueba. Esta división se realizó utilizando la función `train_test_split` de `sklearn.model_selection` con los siguientes parámetros:
- **`test_size=0.2`**: El 20% de los datos se asignó al conjunto de prueba, y el 80% restante al conjunto de entrenamiento.
- **`random_state=42`**: Se utilizó un `random_state` fijo para asegurar la reproducibilidad de la división de los datos, lo que significa que el mismo conjunto de datos se dividirá de la misma manera cada vez que se ejecute el código.

Esto resultó en:
- `X_train_1`, `y_train_1`: Datos para entrenar el modelo.
- `X_test_1`, `y_test_1`: Datos para evaluar el rendimiento del modelo en datos no vistos.

### Entrenamiento y Predicción
El modelo `model_1` fue inicializado como `CatBoostRegressor` y entrenado (`.fit()`) con los datos de entrenamiento `X_train_1` y `y_train_1`. Una vez entrenado, se utilizó para hacer predicciones sobre el conjunto de prueba (`X_test_1`), generando `y_pred_test_1`.

### Métricas de Evaluación
Para evaluar la precisión del modelo, se calcularon las siguientes métricas:
- **Error Cuadrático Medio (MSE)**: `203`
- **Raíz del Error Cuadrático Medio (RMSE)**: `14 minutos`
- **Coeficiente de Determinación (R²)**: `0.8642`

El RMSE de 14 minutos indica que, en promedio, las predicciones del modelo se desvían de los tiempos de vuelo reales en aproximadamente 14 minutos. Un valor de R² de 0.8642 sugiere que el modelo explica aproximadamente el 86.42% de la variabilidad en los tiempos de vuelo, lo que indica un ajuste razonablemente bueno a los datos.

### Uso del Modelo para Predicciones
El modelo entrenado (`model_1`) se puede utilizar para predecir la duración de nuevos vuelos. Para ello, se crea un `DataFrame` (`X_entrada`) con las características de un vuelo específico (Origen, Destino, Aerolinea, periodo_dia, dia_nombre, dia_numero, mes_numero) y se llama al método `.predict()` del modelo. El resultado es la duración estimada del vuelo en minutos. Por ejemplo, la predicción para un vuelo específico fue `111 minutos`.

## Propósito del CSV 'promedio_tiempos_vuelo'

### Subtask:
Explicar cómo se carga el archivo `promedio_tiempos_vuelo.csv` y su rol en la función `demora`, detallando cómo se compara el tiempo de vuelo predicho con un tiempo promedio para determinar si existe una demora significativa.


### Propósito del CSV 'promedio_tiempos_vuelo'

El archivo `promedio_tiempos_vuelo.csv` se carga en el DataFrame `df_tiempo_vuelos`. Este DataFrame contiene los tiempos promedio históricos de vuelos para combinaciones específicas de aerolíneas, orígenes y destinos.

Su rol es fundamental en la función `demora` (o en el bloque de código que realiza esta lógica). Para un vuelo dado (definido por su aerolínea, origen y destino), el código busca en `df_tiempo_vuelos` el `Tiempo_Promedio` correspondiente.

Una vez obtenido el tiempo promedio histórico, se compara con el tiempo de vuelo predicho (`t_vuelo`) por el `model_1`. Se calcula la diferencia absoluta entre el tiempo predicho y el tiempo promedio histórico.

**Criterio de Demora:**

*   Si la diferencia es menor a 15 minutos, se considera que **no hay demora significativa**, y la variable `demora` se establece en `0`.
*   Si la diferencia es igual o mayor a 15 minutos, se considera que **hay una demora significativa**, y la variable `demora` se establece en `1`.

## Identificación de Mejoras y Elementos Sobrantes

### Subtask:
Identificar y explicar el código comentado para 'temporada_alta', la inicialización redundante de variables, y la sección de código para un segundo modelo no implementado.


### Identificación de Mejoras y Elementos Sobrantes

1.  **Cálculo de `temporada_alta` comentado**: Se observa que el código para calcular la columna `temporada_alta` en la celda `GIYzzxM-Mqvx` está completamente comentado. Esto significa que la columna `temporada_alta` fue inicializada a `0` en la celda `LPB1OYSUMhSM` y conserva ese valor para todas las filas, lo que la hace no informativa para el modelo. Debería decidirse si esta característica es relevante y, en caso afirmativo, el código para su cálculo debe ser descomentado y ejecutado. Si no es necesaria, la inicialización y la columna misma podrían ser eliminadas.

2.  **Inicialización redundante de variables**: En la celda `LPB1OYSUMhSM`, las columnas `temporada_alta` y `periodo_dia` se inicializan. Sin embargo, la columna `periodo_dia` es posteriormente sobrescrita en la celda `OQXS0KbrUkwV` con un cálculo más específico. Esto hace que la inicialización de `periodo_dia` sea redundante y podría ser eliminada para simplificar el código.

3.  **Segundo modelo de 'Predicción de Ocupación de Vuelos' incompleto**: En la sección marcada como `### Modelo - Predicción de Ocupación de Vuelos`, la celda `i53Kf5u4pQao` define `X2` y la celda `-m69TUr2Vtiw` define las `categorical_features` para este segundo modelo. Sin embargo, no hay código subsiguiente que entrene o evalúe un modelo (`CatBoostRegressor` o similar) utilizando `X2`. La sección está incompleta y no llega a implementar la predicción de ocupación de vuelos.

Para mejorar la legibilidad, el mantenimiento y la eficiencia del notebook, se recomienda:
*   Descomentar y activar el código de `temporada_alta` si la característica es requerida, o eliminarla si no lo es.
*   Eliminar la inicialización redundante de `periodo_dia`.
*   Completar el desarrollo del segundo modelo para la predicción de ocupación de vuelos, o eliminar la sección si no se va a implementar.

## Final Task

### Subtask:
Resumir las explicaciones, el análisis del modelo y las sugerencias de mejora.


## Summary:

### Data Analysis Key Findings

*   **Objective**: The notebook aims to preprocess flight data, predict flight duration using a machine learning model, and identify potential flight delays by comparing predicted times with historical averages.
*   **Data Preprocessing**:
    *   Flight data from `vuelos_preliminar.csv` is loaded and the last row is removed.
    *   Key date/time columns (`Fecha_Hora_Local_despegue`, `Tiempo_Vuelo`) are converted to appropriate data types (datetime, timedelta).
    *   New features are engineered, including `periodo_dia` (morning, afternoon, night), `dia_nombre`, `dia_numero`, `mes_numero`, and `minutos_vuelo` (flight duration in minutes).
    *   The `temporada_alta` column was initialized to 0, but its calculation code was commented out, leaving it uninformative.
*   **Flight Time Prediction Model (`model_1`)**:
    *   A `CatBoostRegressor` model was trained to predict `minutos_vuelo`.
    *   Input features included `Origen`, `Destino`, `Aerolinea`, `periodo_dia`, `dia_nombre`, `dia_numero`, and `mes_numero`, all treated as categorical.
    *   The model achieved a Mean Squared Error (MSE) of 203, a Root Mean Squared Error (RMSE) of 14 minutes, and an R-squared (R2) of 0.8642, indicating a reasonably good fit.
*   **Delay Detection**:
    *   The `promedio_tiempos_vuelo.csv` file provides historical average flight times.
    *   Predicted flight times are compared against these historical averages. A difference of $\ge$ 15 minutes (absolute value) signifies a significant delay.
*   **Code Redundancies and Incompleteness**:
    *   The `temporada_alta` feature's calculation is commented out, rendering it inactive.
    *   The initialization of `periodo_dia` is redundant as it's immediately overwritten.
    *   A section for a second model (`X2`) intended for "Predicción de Ocupación de Vuelos" is present but incomplete, with no model training or evaluation.

### Insights or Next Steps

*   **Feature Engineering Refinement**: Either activate the commented-out code for `temporada_alta` to make it an informative feature for the model, or remove it along with its initialization to streamline the preprocessing.
*   **Code Clean-up and Completion**: Eliminate redundant variable initializations and either complete the implementation of the second model for flight occupation prediction or remove the incomplete section to improve notebook clarity and maintainability.
